google/cloud/firestore_v1/bulk_batch.py

```diff
@@ -0,0 +1,87 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Helpers for batch requests to the Google Cloud Firestore API."""
+from google.api_core import gapic_v1  # type: ignore
+from google.api_core import retry as retries  # type: ignore
+
+from google.cloud.firestore_v1 import _helpers
+from google.cloud.firestore_v1.base_batch import BaseBatch
+
+
+class BulkWriteBatch(BaseBatch):
+    """Accumulate write operations to be sent in a batch. Use this over `WriteBatch`
+    for higher volumes (e.g., via `BulkWriter`) and when the order of operations
+    within a given batch is unimportant.
+
+    Because the order in which individual write operations land in the database
+    is not guaranteed, `batch_write` RPCs can never contain multiple operations
+    to the same document. If calling code detects a second write operation to a
+    known document reference, it should first cut off the previous batch and
+    send it, then create a new batch starting with the latest write operation.
+    In practice, the [Async]BulkWriter classes handle this.
+
+    This has the same set of methods for write operations that
+    :class:`~google.cloud.firestore_v1.document.DocumentReference` does,
+    e.g. :meth:`~google.cloud.firestore_v1.document.DocumentReference.create`.
+
+    Args:
+        client (:class:`~google.cloud.firestore_v1.client.Client`):
+            The client that created this batch.
+    """
+
+    def __init__(self, client) -> None:
+        super(BulkWriteBatch, self).__init__(client=client)
+
+    def commit(
+        self, retry: retries.Retry = gapic_v1.method.DEFAULT, timeout: float = None
+    ):
+        """Writes the changes accumulated in this batch.
+
+        Write operations are not guaranteed to be applied in order and must not
+        contain multiple writes to any given document. Preferred over `commit`
+        for performance reasons if these conditions are acceptable.
+
+        Args:
+            retry (google.api_core.retry.Retry): Designation of what errors, if any,
+                should be retried.  Defaults to a system-specified policy.
+            timeout (float): The timeout for this request.  Defaults to a
+                system-specified value.
+
+        Returns:
+            List[:class:`google.cloud.proto.firestore.v1.write.BatchWriteResult`, ...]:
+            The write results corresponding to the changes committed, returned
+            in the same order as the changes were applied to this batch. A
+            write result contains an ``update_time`` field.
+        """
+        request, kwargs = self._prep_commit(retry, timeout)
+
+        save_response = self._client._firestore_api.batch_write(
+            request=request, metadata=self._client._rpc_metadata, **kwargs,
+        )
+
+        self._write_pbs = []
+        self.write_results = results = list(save_response.write_results)
```

Comment:
Is this intermediate assignment to results to manage some sort of race? If so, is there not also an issue with commit prepping and setting `_write_pbs` to empty?

Reply:
I'm not aware of any possibly race conditions here, but if you think there is we should chat more.

Also worth noting that this functionality is more or less copied out of the older `batch.py` file.